[1 - 2단계 방탈출 예약 대기] 미미(홍향미) 미션 제출합니다.

hongkikii opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [X] 미션의 필수 요구사항을 모두 구현했나요?
- [X] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [X] 애플리케이션이 정상적으로 실행되나요?
- [X] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
[넵!](https://prolog.techcourse.co.kr/studylogs/4354)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
- [x] 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.
- [x] else 예약어를 쓰지 않는다.
- [ ] 모든 원시 값과 문자열을 포장한다.
➡️ 특정 로직을 포함하지 않는다면 포장하지 않았습니다.

- [ ] 일급 컬렉션을 쓴다.
➡️ List가 쓰인 응답 객체를 포장하지 않았습니다. 당장 포장의 장점을 느끼지 못하였습니다!

- [ ] 한 줄에 점을 하나만 찍는다.
➡️ 한 줄에 점을 하나만 찍었을 때 오히려 가독성을 해친다고 생각되는 경우 이를 지키지 않았습니다.

- [X] 줄여 쓰지 않는다(축약 금지).
- [X] 모든 엔티티를 작게 유지한다.

- [ ] 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
➡️ 도메인 객체와 DB의 연동이 쉽도록 하기 위해 지키지 못하였습니다.

- [ ] getter/setter/프로퍼티를 쓰지 않는다.
➡️ dto <-> 엔티티 간 변환 과정에서 getter를 사용하였습니다.

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [X] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

안녕하세요, 범블비!
레벨1 블랙잭 미션에 이어 다시 만나뵙게 되어 반갑습니다 😃

이전에 JPA는 기본 사용법을 익히고 따라해보는 정도로 프로젝트에서 사용해봤어요.
그래서 이번 미션을 진행하면서는 학습 키워드를 깊게 이해하려고 노력했습니다.
- JPA와 ORM이 무엇인지
- JPA가 주는 이점이 무엇인지
- 연관관계 매핑에서 mappedBy의 역할은 무엇인지를 중심으로 고민했습니다.

그런데 프롤로그 질문 중에 'JPA를 사용하면서 엔티티 설계를 더욱 객체지향적으로 할 수 있었는지?'라는 질문에 대해서는 
아직 그런 경험이나 이해가 부족해서 아쉬웠습니다..
앞으로 리뷰를 포함한 남은 미션을 진행한 후에는 이 질문에 대해 '그렇다!'라고 말할 수 있게 되고 싶습니다!

또 ArgumentResolver, HandlerInterceptor에 대한 테스트 코드와 RestAssured를 이용한 통합 테스트 코드를 작성했어요.
기존 페어의 코드에는 domain과 (repository를 사용한) service 코드에 대해서만 테스트 코드가 작성이 되어 있었는데요.

토큰이 잘 생성되는지, 권한을 잘 체크하는지 확인하고자 ArgumentResolver, HandlerInterceptor에 대한 테스트 코드를 추가해줬어요.
그리고 '토큰 생성 및 권한 체크 -> 컨트롤러 호출 -> 비즈니스 로직 -> 적절한 형식의 응답'이라는 전반적인 과정이 잘 수행되는지 체크하기 위해 통합 테스트 코드를 추가했어요.

통합 테스트를 작성하면서 궁금했던 것이 있었습니다.
제가 작성한 통합 테스트의 목적은 '내가 작성한 코드 외에 외부 통신까지 잘 이루어지고 있는지를 확인하기 위함'이라고 생각해요.
그렇다면 과연 모든 요구사항(시나리오)에 대한 통합 테스트 코드가 필요할까? 라는 의문이 있었던 것 같습니다.
물론 모두 작성한다면 좋겠지만, 미션을 진행하면서 요구사항이 수정됨에 따라 테스트를 작성(수정)하는 게 비용이 될 수 있다는 걸 점점 느끼고 있습니다.. 그렇다면 효율적으로 테스트를 작성하는 게 중요할 수 있겠다는 생각이 들었어요!

리뷰 잘 부탁드리겠습니다 🙇🏻‍♀️
감사합니다!

++) 변경 커밋 [링크](https://github.com/woowacourse/spring-roomescape-waiting/pull/241/files/e236712301a413f47c55c558b49a4d73e4749b12..171dfb48f08c033f5480265678ad155f183d72cd)를 안올려서 추가합니다..! 🙏🏻



### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

eoehd1ek committed:
chore: 이전 미션 코드 가져오기

eoehd1ek committed:
docs: 1단계 요구 사항 추가

eoehd1ek committed:
chore: gradle 의존성 추가

eoehd1ek committed:
feat 엔티티 매핑 추가

hongkikii committed:
feat: 연관관계 매핑 추가

hongkikii committed:
feat: JpaRepository 사용

eoehd1ek committed:
docs: 2단계 요구 사항 추가

hongkikii committed:
feat: 내 예약 목록 조회 기능 추가

eoehd1ek committed:
docs: 리팩터링 목록 추가

hongkikii committed:
refactor: ArgumentResolver 어노테이션 체크 추가

hongkikii committed:
test: CheckAdminInterceptor 권한 확인 테스트 추가

eoehd1ek committed:
test: Theme API 통합 테스트 작성

hongkikii committed:
test: 통합 테스트 추가

eoehd1ek committed:
test: 누락된 통합 테스트 추가

hongkikii committed:
chore: 패키지 구조 변경

ddaaac commented on src/main/java/roomescape/business/domain/Member.java:
```diff
@@ -16,14 +24,21 @@ public Member(final Long id, final String name, final String role, final String
         this.role = null;
         this.email = null;
         this.password = null;
     }
```
이 생성자는 왜 필요한걸까요? 좀 어색해보여요.

hongkikii replied:
Reservation 도메인 테스트 시 객체를 쉽게 생성하기 위해 만들었는데,
필드 값을 null로 허용하는 생성자를 둘 경우 당장의 편리함보다 프로덕션 코드에서의 위험성이 더 클 것 같아 삭제하였습니다.

테스트에 정말 필요하다면 테스트 픽스처를 사용하거나 정적 팩터리 메서드로 테스트용이라는 이름을 주는 게 훨씬 더 좋겠네요!

ddaaac commented on src/main/java/roomescape/business/domain/Reservation.java:
```diff
@@ -1,29 +1,43 @@
+    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private LocalDate date;
+    @ManyToOne
```
`@ManyToOne`의 기본 fetch 전략이 무엇이고, eager와 lazy 로딩의 차이에 대해서 학습해보시면 좋을 것 같아요.

hongkikii replied:
@xxxToOne은 기본 fetch 전략이 즉시 로딩이니 지연 로딩으로 설정을 바꿔서 불필요한 성능 저하를 없애고,
필요한 경우 fetch join이나 @EntityGraph를 사용해주는 것이 좋겠네요! 👍🏻

ddaaac commented on src/main/java/roomescape/business/domain/Member.java:
```diff
@@ -1,12 +1,20 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private String name;
```
`@Column`을 통해 스키마에도 제약 조건들을 명시해주시면 좋을 것 같네요.

hongkikii replied:
그러네요! 기존 schema.sql에서 작성했던 부분들을 명시해줘야겠네요!

ddaaac commented on src/main/java/roomescape/business/domain/Member.java:
```diff
@@ -1,12 +1,20 @@
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+
+@Entity
 public class Member {
```
`@Table`을 통해 테이블 이름을 명시해주세요.

hongkikii replied:
지금 당장은 아무 생각 없이 애플리케이션을 실행할 때마다 테이블을 자동으로 생성해주고 있어서 의식하지 못 했는데,
만약 프로젝트에서 이미 설정한 DB가 있다면 JPA가 자동으로 설정해주는 이름과 다를 수 있기 때문에 직접 명시를 해주는 것이 좋겠네요!
Jpa 구현체나 직접 설정한 네이밍 전략에 의존하지 않게 할 수 있구요.
코드로 명확하게 테이블의 이름을 나타낼 수 있어서 가독성 측면에서도 좋은 것 같습니다 👍🏻

ddaaac commented on src/main/java/roomescape/business/domain/ReservationTime.java:
```diff
@@ -0,0 +1,52 @@
+import jakarta.persistence.Id;
+import java.time.LocalTime;
+
+@Entity
+public class ReservationTime {
```
> 그런데 프롤로그 질문 중에 'JPA를 사용하면서 엔티티 설계를 더욱 객체지향적으로 할 수 있었는지?'라는 질문에 대해서는
아직 그런 경험이나 이해가 부족해서 아쉬웠습니다..

개인적으로도 JPA와 같은 기술을 사용하면서 "더욱 객체지향적"이라고 얘기하는 건 잘 모르겠네요.
설계를 완전 객체지향적으로 하고 싶다면 도메인과 DB 엔티티를 분리하는 게 맞지 않나 싶기도 하구요.

여튼, 우리가 다루는 도메인은 대부분 DB에 영속화가 필요하고 그러면서도 객체지향이라는 방식을 지키고 싶어하기 때문에 JPA를 활용하는 거라고 생각하고 접근하시면 좋을 것 같습니다. 조금만 신경쓰고 학습해서 사용한다면 쿼리 같은 인프라적인 고려요소를 크게 신경쓰지 않고도 객체지향적인 구현을 할 수 있으니깐요.

hongkikii replied:
JPA라는 기술 자체가 객체지향적인 설계를 돕는다고 할 순 없겠지만,
JPA 덕분에 객체지향에 조금 더 쉽게 집중할 수 있겠네요..! 👍🏻

ddaaac commented on src/main/java/roomescape/business/service/ReservationService.java:
```diff
@@ -6,33 +6,34 @@
 @Service
 public class ReservationService {
 
-    private final ReservationDao reservationDao;
```
`@Transactional` 에 대해서 학습해봐도 좋을 것 같아요.

hongkikii replied:
찾아보니 @Transactional을 설정하지 않으면 여러 문제가 발생하는데,
치명적으로 flush가 (직접 작성해주지 않는 이상) 되지 않고 에외 시 롤백시키지 않는다는 문제가 있네요.
어노테이션 추가하겠습니다!

ddaaac commented on src/test/java/roomescape/business/service/AuthServiceTest.java:
```diff
@@ -6,23 +6,20 @@
+@DataJpaTest
 @Sql("classpath:data-authService.sql")
 class AuthServiceTest {
 
     private final AuthService authService;
 
     @Autowired
-    public AuthServiceTest(final JdbcTemplate jdbcTemplate) {
-        final MemberDao memberDao = new JdbcMemberDao(jdbcTemplate);
-        this.authService = new AuthService(memberDao);
+    public AuthServiceTest(final MemberRepository memberRepository) {
+        this.authService = new AuthService(memberRepository);
```
하나 질문을 던져보자면 `DataJpaTest`는 통합테스트일까요?

hongkikii replied:
@DataJpaTest는 영속성 계층의 테스트를 위한 것으로 알고 있습니다.
따라서 스프링 컨텍스트 내에서 Jpa Repository와 관련된 계층만 사용하는 슬라이스 테스트로 알고 있는데요.

만약 @DataJpaTest가 자동으로 설정해주는 Bean의 범위 내에서 테스트를 진행한다면 통합 테스트가 아니라고 생각합니다.
(통합 테스트는 애플리케이션에서 의존하는 다른 요소들을 실제로 사용하는 것이라고 생각하는데, 영속성 계층 외의 Bean을 설정해주지 않기 때문입니다!)

다만, 저와 페어는 service 테스트 코드에서 service 클래스를 직접 생성하고 @DataJpaTest를 사용하였는데요.
service 클래스에서 의존하는 Repository를 Fake나 Mock 처리 하지 않고, DB와 상호작용하고 있는지까지 테스트하기 위해 @DataJpaTest를 사용하였습니다.
따라서 저희 코드에서는 통합 테스트를 위해 @DataJpaTest가 사용됐다고 할 수 있을 것 같습니다.

근데 질문 주신 것에 대한 답변을 정리하다보니, @DataJpaTest 어노테이션을 목적과는 다르게 사용했다는 생각이 들기도 하네요..
필요한 Bean만 사용하기 위해 선택한 방식인데, 혹시 적절하지 않은 방법일까요?!

ddaaac replied:
지금 테스트는 서비스 레이어 + 영속 레이어 를 함께 테스트하고 있습니다. 따라서 통합 테스트로 분리하는 게 맞겠죠.

또, 지금이야 서비스 레이어가 영속 레이어에만 의존하고 있지만 일반적으로 서비스 레이어는 다른 서비스 레이어에도 의존합니다. 아니면 영속 레이어가 아닌 외부 클라이언트에 의존할 수 있구요. 따라서 `@DataJpaTest`만으로 필요한 빈들을 불러오지 못하는 경우도 많이 생기구요.

`@DataJpaTest`는 말 그대로 영속 레이어만 테스트할 때 사용해주시면 될 것 같습니다.

ddaaac commented on src/test/java/roomescape/presentation/AdminControllerIntegrationTest.java:
```diff
@@ -0,0 +1,292 @@
+import roomescape.presentation.dto.LoginRequest;
+import roomescape.presentation.dto.ReservationRequest;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+class AdminControllerIntegrationTest {
```
> 그렇다면 과연 모든 요구사항(시나리오)에 대한 통합 테스트 코드가 필요할까? 라는 의문이 있었던 것 같습니다.


맞습니다. 사실 모든 요구사항에 대해서 테스트를 만드는 건 정말 힘든 일이고, 추후 관리도 안 될 가능성이 크다고(거의 100%라고 생각해요) 생각합니다. 그래서 적당히 핵심 로직에 대해서만 테스트하는 게 필요하다고 생각합니다.

다만, 어떤 로직을 테스트해야하고 넘겨도 될지를 판단하는 건 쉽지 않습니다. 테스트를 작성하지 않은 로직이 나중에 문제가 될 수도 있거든요. 가능하면 최대한 많은 테스트를 작성해보면서 미미만의 방식을 찾으면 좋겠네요.

hongkikii replied:
답변 감사합니다 !!
지금은 배우는 단계이니 최대한 많이, 다양한 방식으로 테스트를 작성해보면서 저만의 직관을 키워봐야겠네요 🤔

ddaaac commented on src/test/java/roomescape/config/CheckAdminInterceptorTest.java:
```diff
@@ -0,0 +1,102 @@
+import roomescape.exception.UnauthorizedException;
+import roomescape.presentation.dto.LoginMember;
+
+@WebMvcTest(CheckAdminInterceptor.class)
+class CheckAdminInterceptorTest {
```
다양한 방식으로 테스트들 정말 잘 구현해주셨습니다 👍

ddaaac reviewed (changes requested):
안녕하세요 미미 👋 

구현 잘 해주셨네요.

몇 가지 코멘트 남겼으니 확인해주세요.

hongkikii committed:
refactor: 테이블 이름 명시

hongkikii committed:
refactor: Column 제약 조건 명시

hongkikii committed:
refactor: @ManyToOne fetch 전략 수정

hongkikii committed:
style: 메서드 순서 변경

hongkikii committed:
style: 메서드 순서 변경

hongkikii committed:
refactor: @JoinColumn 명시

hongkikii committed:
refactor: @Transactional 설정 추가

hongkikii committed:
style: 메서드 네이밍 수정

hongkikii committed:
style: 개행 추가 및 불필요 코드 제거

hongkikii committed:
fix: 테스트 코드 오류 수정

- 엔티티 객체 응답 값으로 사용 시 지연 로딩으로 인해 null값 발생하여 응답 전용 dto 생성
- test용 sql과 엔티티 이름 충돌 막기 위해 기존 테이블 네이밍으로 되돌림

hongkikii committed:
refactor: 불필요 생성자 제거

hongkikii commented:
안녕하세요, 범블비!
말씀해 주신 부분 따라 고민해보고 코드 수정하였습니다.
이번 미션에서 추가로 학습해야 할 키워드들을 명시해주셔서 놓쳤던 부분을 챙길 수 있었습니다 👍🏻

남은 리뷰도 잘 부탁드립니다 🙏🏻 
감사합니다 :)

ddaaac commented on src/main/java/roomescape/config/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,52 @@
+    ) throws UnauthorizedException {
+        final HttpServletRequest request = (HttpServletRequest) webRequest.getNativeRequest();
+        final String accessToken = getAccessToken(request);
+        return authService.getLoginMemberByAccessToken(accessToken);
```
사실 이런거 보면 좀 고민이 되긴합니다. 인터셉터와 완전 동일한 정보가 필요한데 굳이 db를 두번 왔다갔다 할 필요가 있을까 싶기도 하구요.

한 번쯤 가볍게 고민해보셔도 좋을 것 같습니다.

ddaaac reviewed (approved):
안녕하세요 미미 👋 

리뷰 반영 잘 해주셨습니다.

다음 단계 진행해주세요!
